mesonbuild/environment.py

```diff
@@ -693,28 +705,39 @@ def detect_static_linker(self, compiler):
             linker = self.cross_info.config['binaries']['ar']
             if isinstance(linker, str):
                 linker = [linker]
+            linkers = [linker]
         else:
             evar = 'AR'
             if evar in os.environ:
-                linker = shlex.split(os.environ[evar])
+                linkers = [shlex.split(os.environ[evar])]
             elif isinstance(compiler, VisualStudioCCompiler):
-                linker = [self.vs_static_linker]
+                linkers = [self.vs_static_linker]
+            elif isinstance(compiler, GnuCompiler):
+                # Use gcc-ar if available; needed for LTO
+                linkers = [self.gcc_static_linker, self.default_static_linker]
+            elif isinstance(compiler, ClangCompiler):
+                # Use llvm-ar if available; needed for LTO
+                linkers = [self.clang_static_linker, self.default_static_linker]
             else:
-                linker = [self.default_static_linker]
-        if 'lib' in linker or 'lib.exe' in linker:
-            arg = '/?'
-        else:
-            arg = '--version'
-        try:
-            p, out, err = Popen_safe(linker + [arg])
-        except OSError:
-            raise EnvironmentException('Could not execute static linker "%s".' % ' '.join(linker))
-        if '/OUT:' in out or '/OUT:' in err:
-            return VisualStudioLinker(linker)
-        if p.returncode == 0:
-            return ArLinker(linker)
-        if p.returncode == 1 and err.startswith('usage'): # OSX
-            return ArLinker(linker)
+                linkers = [self.default_static_linker]
+        popen_exceptions = {}
+        for linker in linkers:
+            if 'lib' in linker or 'lib.exe' in linker:
+                arg = '/?'
+            else:
+                arg = '--version'
+            try:
+                p, out, err = Popen_safe(linker + [arg])
+            except OSError as e:
+                popen_exceptions[' '.join(linker + [arg])] = e
+                continue
+            if '/OUT:' in out or '/OUT:' in err:
+                return VisualStudioLinker(linker)
```

Comment:
__[flake8]__
* [F405] 'VisualStudioLinker' may be undefined, or defined from star imports: .compilers

<!-- Comment by [SideCI](https://sideci.com) -->